README
NAME
    Package::Debug - Add ENV/Config controlled debug tokens to your code

VERSION
    version 0.2.1

SYNOPSIS
    There is a lot of code on "CPAN" that has something like this in it:

        our $DEBUG = $ENV{MY_PACKAGE_NAME_DEBUG};

    or something like

        sub DEBUG {
            return if not $ENV{MY_PACKAGE_NAME_DEBUG};
            <real debug code>
        }

    or something like:

        if( $ENV{MY_PACKAGE_NAME_DEBUG} ) {
            *DEBUG=sub{ withdebug }
        } else {
            *DEBUG=sub { noop }
        }

    These are mostly simple and straight forward, ... however, they
    artificially limit what you can do, at the cost of making ugly code.

    This module aims to implement the common utility, with less fuss:

        $ENV{MY_BAZ_DEBUG} = 1;

        package My::Baz;

        use Package::Debug;

        ...

        sub foo {
            DEBUG("message");
        }

    And all the right things should still occur.

    Additionally, this module will eventually add a bunch of features, that
    are by default off, but can be toggled on using environment or
    configuration files.

EXPECTED FEATURES
    *   Deferrable debug mechanism

        The defacto "DEBUG()" stub when not in a debug environment should be
        a no-op, or as close to a no-op as possible.

        However, when debugging is turned on, debugging back ends should
        also be controllable via env/configuaration, and proxy to things
        like Log::Message and friends.

    *   Per-package debug granularity

        Every package will get its own independent DEBUG key, and DEBUG for
        a class can be toggled with an %ENV key relevant to that class.

    *   Global Debugging

        In addition to package level granularity, global debugging can also
        be enabled, while still seeing the individual packages the debug
        message emanates from.

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Package-Debug

{{$NEXT}}

0.2.2 2013-09-26T06:54:05Z
 [00 Minor]
 - Rebuild to get latest Test::Compile to maybe avoid some Win32 bugs.

 [Dependencies::Added / test requires]
 - File::Spec

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.8.2 → v1.8.3

 [Dependencies::Changed / runtime requires]
 - Package::Stash 0.35 → 0.36

 [Dependencies::Changed / test requires]
 - Test::Fatal 0.010 → 0.013

 [!Major]
 - Noisiness of output during general usage reduced.
 - Additional sanity checks in place to warn if anything that *would* be loaded via bootstrap, is unable to be bootstrapped due to wrong plugin order.
 - Tests regenerated with newer toolkit.

 [Dependencies::Added / develop requires]
 - Test::CPAN::Changes 0.19
 - Test::Kwalitee 1.08

 [Dependencies::Added / runtime requires]
 - lib

 [Dependencies::Added / test requires]
 - IO::Handle
 - IPC::Open3

 [Dependencies::Changed / build requires]
 - Module::Build 0.4004 → 0.4007

 [Dependencies::Changed / configure requires]
 - Module::Build 0.4004 → 0.4007

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.7.2 → v1.8.3

 [Dependencies::Removed / test requires]
 - File::Find
 - File::Temp

0.2.1 2013-08-30T21:04:58Z
 [!Minor]
 - Bump to get new Test::Compile
 - Increase deps on Test-essential things to try rule out weird silent
   failures on a few targets

 [Dependencies::Added / develop requires]
 - Test::Kwalitee 1.08

 [Dependencies::Added / test requires]
 - IO::Handle
 - IPC::Open3

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.8.0 → v1.8.2

 [Dependencies::Changed / runtime requires]
 - Package::Stash 0 → 0.35

 [Dependencies::Changed / test requires]
 - Test::Fatal 0 → 0.010
 - Test::Output 0 → 1.02

 [Dependencies::Removed / test requires]
 - Capture::Tiny
 - blib

0.2.0 2013-08-09T22:47:30Z
 [!Major]
 - runtime_switchable feature removed in favour of more general purpose
   code, that is performant even *with* runtime switchability being on by
   default.
 - Class methods starting with full_ removed as they're no longer used.
 - Additionally, loading `Readonly` to implement above feaures usually
   costed more than it saved.
 - Internals changes now means that only ~90 statements are required to
   execute during ->import() whereas previously it was more like ~500
   statements.

 [Dependencies::Added / runtime requires]
 - Package::Stash

 [Dependencies::Added / test requires]
 - Capture::Tiny
 - blib

 [Dependencies::Changed / build requires]
 - Module::Build 0.4005 → 0.4007

 [Dependencies::Changed / configure requires]
 - Module::Build 0.4005 → 0.4007

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.7.4 → v1.8.0

 [Dependencies::Removed / runtime requires]
 - B
 - Readonly
 - Scalar::Util

 [Dependencies::Removed / test requires]
 - File::Find
 - File::Temp

 [Internals]
 - Package stash mangling now done by Package::Stash
 - '_has' attribute generator removed in favor of hand-inlined methods,
   which is easier to profile and debug.

0.1.0 2013-07-31T03:19:56Z
 - First version.



lib/Package/Debug.pm
use strict;
use warnings;

package Package::Debug;
BEGIN {
  $Package::Debug::AUTHORITY = 'cpan:KENTNL';
}
{
  $Package::Debug::VERSION = '0.2.1';
}

# ABSTRACT: Add ENV/Config controlled debug tokens to your code


sub import {
  my ( $self, %args ) = @_;
  require Package::Debug::Object;
  my $object = Package::Debug::Object->new(%args);
  $object->auto_set_into(1);
  $object->inject_debug_value();
  $object->inject_debug_sub();
  return $object;
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Package::Debug - Add ENV/Config controlled debug tokens to your code

=head1 VERSION

version 0.2.1

=head1 SYNOPSIS

There is a lot of code on C<CPAN> that has something like this in it:

    our $DEBUG = $ENV{MY_PACKAGE_NAME_DEBUG};

or something like

    sub DEBUG {
        return if not $ENV{MY_PACKAGE_NAME_DEBUG};
        <real debug code>
    }

or something like:

    if( $ENV{MY_PACKAGE_NAME_DEBUG} ) {
        *DEBUG=sub{ withdebug }
    } else {
        *DEBUG=sub { noop }
    }

These are mostly simple and straight forward, ... however, they artificially limit what you can do,
at the cost of making ugly code.

This module aims to implement the common utility, with less fuss:

    $ENV{MY_BAZ_DEBUG} = 1;

    package My::Baz;

    use Package::Debug;

    ...

    sub foo {
        DEBUG("message");
    }

And all the right things should still occur.

Additionally, this module will eventually add a bunch of features, that are by default off, but can be toggled
on using environment or configuration files.

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Package::Debug",
    "interface":"exporter"
}


=end MetaPOD::JSON

=head1 EXPECTED FEATURES

=over 4

=item * Deferrable debug mechanism

The defacto C<DEBUG()> stub when not in a debug environment should be a no-op, or as close to a no-op as possible.

However, when debugging is turned on, debugging back ends should also be controllable via env/configuaration,
and proxy to things like Log::Message and friends.

=item * Per-package debug granularity

Every package will get its own independent DEBUG key, and DEBUG for a class can be toggled
with an C<%ENV> key relevant to that class.

=item * Global Debugging

In addition to package level granularity, global debugging can also be enabled, while still seeing the individual packages the debug message emanates from.

=back

=head1 AUTHOR

Kent Fredric <kentfredric@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Package/Debug/Object.pm
use strict;
use warnings;

package Package::Debug::Object;
BEGIN {
  $Package::Debug::Object::AUTHORITY = 'cpan:KENTNL';
}
{
  $Package::Debug::Object::VERSION = '0.2.1';
}

# ABSTRACT: Object oriented guts to Package::Debug


my %env_key_styles = ( default => 'env_key_from_package', );


my %env_key_prefix_styles = ( default => 'env_key_prefix_from_package', );


my %log_prefix_styles = (
  short => 'log_prefix_from_package_short',
  long  => 'log_prefix_from_package_long',
);


my %debug_styles = (
  'prefixed_lines' => 'debug_prefixed_lines',
  'verbatim'       => 'debug_verbtaim',
);


sub new {
  my ( $self, %args ) = @_;
  return bless \%args, $self;
}


sub debug_style {
  return $_[0]->{debug_style} if exists $_[0]->{debug_style};
  return ( $_[0]->{debug_style} = 'prefixed_lines' );
}

sub set_debug_style {
  $_[0]->{debug_style} = $_[1];
  return $_[0];
}


sub env_key_aliases {
  return $_[0]->{env_key_aliases} if exists $_[0]->{env_key_aliases};
  return ( $_[0]->{env_key_aliases} = [] );
}

sub set_env_key_aliases {
  $_[0]->{env_key_aliases} = $_[1];
  return $_[0];
}


sub env_key_prefix_style {
  return $_[0]->{env_key_prefix_style} if exists $_[0]->{env_key_prefix_style};
  return ( $_[0]->{env_key_prefix_style} = 'default' );
}

sub set_env_key_prefix_style {
  $_[0]->{env_key_prefix_style} = $_[1];
  return $_[0];
}


sub env_key_style {
  return $_[0]->{env_key_style} if exists $_[0]->{env_key_style};
  return ( $_[0]->{env_key_style} = 'default' );
}

sub set_env_key_style {
  $_[0]->{env_key_style} = $_[1];
  return $_[0];
}


sub into {
  return $_[0]->{into} if exists $_[0]->{into};
  die 'Cannot vivify ->into automatically, pass to constructor or ->set_into() or ->auto_set_into()';
}

sub set_into {
  $_[0]->{into} = $_[1];
  return $_[0];
}


sub into_level {
  return $_[0]->{into_level} if exists $_[0]->{into_level};
  return ( $_[0]->{into_level} = 0 );
}

sub set_into_level {
  $_[0]->{into_level} = $_[1];
  return $_[0];
}


sub sub_name {
  return $_[0]->{sub_name} if exists $_[0]->{sub_name};
  return ( $_[0]->{sub_name} = 'DEBUG' );
}

sub set_sub_name {
  $_[0]->{sub_name} = $_[1];
  return $_[0];
}


sub value_name {
  return $_[0]->{value_name} if exists $_[0]->{value_name};
  return ( $_[0]->{value_name} = 'DEBUG' );
}

sub set_value_name {
  $_[0]->{value_name} = $_[1];
  return $_[0];
}


sub env_key {
  return $_[0]->{env_key} if exists $_[0]->{env_key};
  my $style = $_[0]->env_key_style;
  if ( not exists $env_key_styles{$style} ) {
    die "No such env_key_style $style, options are @{ keys %env_key_styles }";
  }
  my $method = $env_key_styles{$style};
  return ( $_[0]->{env_key} = $_[0]->$method() );
}

sub set_env_key {
  $_[0]->{env_key} = $_[1];
  return $_[0];
}


sub env_key_prefix {
  return $_[0]->{env_key_prefix} if exists $_[0]->{env_key_prefix};
  my $style = $_[0]->env_key_prefix_style;
  if ( not exists $env_key_prefix_styles{$style} ) {
    die "No such env_key_prefix_style $style, options are @{ keys %env_key_prefix_styles }";
  }
  my $method = $env_key_prefix_styles{$style};
  return ( $_[0]->{env_key_prefix} = $_[0]->$method() );
}

sub set_env_key_prefix {
  $_[0]->{env_key_prefix} = $_[1];
  return $_[0];
}


sub debug_sub {
  return $_[0]->{debug_sub} if exists $_[0]->{debug_sub};
  my $style = $_[0]->debug_style;
  if ( not exists $debug_styles{$style} ) {
    die "No such debug_style $style, options are @{ keys %debug_styles }";
  }
  my $method = $debug_styles{$style};
  return ( $_[0]->{debug_sub} = $_[0]->$method() );
}

sub set_debug_sub {
  $_[0]->{debug_sub} = $_[1];
  return $_[0];
}


sub log_prefix_style {
  return $_[0]->{log_prefix_style} if exists $_[0]->{log_prefix_style};
  my $style = 'short';
  $style = $ENV{PACKAGE_DEBUG_LOG_PREFIX_STYLE} if $ENV{PACKAGE_DEBUG_LOG_PREFIX_STYLE};
  return ( $_[0]->{log_prefix_style} = $style );
}

sub set_log_prefix_style {
  $_[0]->{log_prefix_style} = $_[1];
  return $_[0];
}


sub log_prefix {
  return $_[0]->{log_prefix} if exists $_[0]->{log_prefix};
  my $style = $_[0]->log_prefix_style;
  if ( not exists $log_prefix_styles{$style} ) {
    die "Unknown prefix style $style, should be one of @{ keys %log_prefix_styles }";
  }
  my $method = $log_prefix_styles{$style};
  return ( $_[0]->{log_prefix} = $_[0]->$method() );
}

sub set_log_prefix {
  $_[0]->{log_prefix} = $_[1];
  return $_[0];
}


sub is_env_debugging {
  return $_[0]->{is_env_debugging} if exists $_[0]->{is_env_debugging};
  if ( $ENV{PACKAGE_DEBUG_ALL} ) {
    return ( $_[0]->{is_env_debugging} = 1 );
  }
  for my $key ( $_[0]->env_key, @{ $_[0]->env_key_aliases } ) {
    next unless exists $ENV{$key};
    next unless $ENV{$key};
    return ( $_[0]->{is_env_debugging} = 1 );
  }
  return ( $_[0]->{is_env_debugging} = 0 );
}

sub set_is_env_debugging {
  $_[0]->{is_env_debugging} = $_[1];
  return $_[0];
}


sub into_stash {
  return $_[0]->{into_stash} if exists $_[0]->{into_stash};
  require Package::Stash;
  return ( $_[0]->{into_stash} = Package::Stash->new( $_[0]->into ) );
}

sub set_into_stash {
  $_[0]->{into_stash} = $_[1];
  return $_[0];
}


sub auto_set_into {
  my ( $self, $add ) = @_;
  $_[0]->{into} = [ caller( $self->into_level + $add ) ]->[0];
  return $self;
}


# Note: Heavy hand-optimisation going on here, this is the hotpath
sub debug_prefixed_lines {
  my $self   = shift;
  my $prefix = $self->log_prefix;
  return sub {
    my (@message) = @_;
    for my $line (@message) {
      *STDERR->print( '[' . $prefix . '] ' ) if defined $prefix;
      *STDERR->print($line);
      *STDERR->print("\n");
    }
  };
}


sub debug_verbatim {
  my $self = shift;
  return sub {
    *STDERR->print(@_);
  };
}


sub env_key_from_package {
  return $_[0]->env_key_prefix() . '_DEBUG';
}


sub env_key_prefix_from_package {
  my $package = $_[0]->into;
  $package =~ s{
    ::
  }{_}msxg;
  return uc $package;
}


sub log_prefix_from_package_short {
  my $package = $_[0]->into;
  if ( ( length $package ) < 10 ) {
    return $package;
  }
  my (@tokens) = split /::/msx, $package;
  my ($suffix) = pop @tokens;
  for (@tokens) {
    if ( $_ =~ /[[:upper:]]/msx ) {
      $_ =~ s/[[:lower:]]+//msxg;
      next;
    }
    $_ = substr $_, 0, 1;
  }
  my ($prefix) = join q{:}, @tokens;
  return $prefix . q{::} . $suffix;
}


sub log_prefix_from_package_long {
  return $_[0]->into;
}


sub inject_debug_value {
  my $value_name = $_[0]->value_name;
  return if not defined $value_name;
  my $value = $_[0]->is_env_debugging;
  my $stash = $_[0]->into_stash;
  if ( $stash->has_symbol( q[$] . $value_name ) ) {
    $value = $stash->get_symbol( q[$] . $value_name );
    $stash->remove_symbol( q[$] . $value_name );
  }
  $stash->add_symbol( q[$] . $value_name, \$value );
  return $_[0];
}

sub _wrap_debug_sub {
  my $sub_name = $_[0]->sub_name;
  return if not defined $sub_name;
  my $value_name       = $_[0]->value_name;
  my $is_env_debugging = $_[0]->is_env_debugging;
  if ( not defined $value_name and not $is_env_debugging ) {
    return sub { };
  }
  my $real_debug = $_[0]->debug_sub;
  my $symbol     = $_[0]->into_stash->get_symbol( q[$] . $value_name );
  return sub {
    return unless ${$symbol};
    goto $real_debug;
  };
}


sub inject_debug_sub {
  $_[0]->into_stash->add_symbol( q[&] . $_[0]->sub_name, $_[0]->_wrap_debug_sub );
  return $_[0];
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Package::Debug::Object - Object oriented guts to Package::Debug

=head1 VERSION

version 0.2.1

=head1 METHODS

=head2 C<new>

    my $object = Package::Debug::Object->new(%args);

=head2 C<debug_style>

=head2 C<set_debug_style>

=head2 C<env_key_aliases>

=head2 C<set_env_key_aliases>

=head2 C<env_key_prefix_style>

=head2 C<set_env_key_prefix_style>

=head2 C<env_key_style>

=head2 C<set_env_key_style>

=head2 C<into>

=head2 C<set_into>

=head2 C<into_level>

=head2 C<set_into_level>

=head2 C<sub_name>

=head2 C<set_sub_name>

=head2 C<value_name>

=head2 C<set_value_name>

=head2 C<env_key>

=head2 C<set_env_key>

=head2 C<env_key_prefix>

=head2 C<set_env_key_prefix>

=head2 C<debug_sub>

=head2 C<set_debug_sub>

=head2 C<log_prefix_style>

=head2 C<set_log_prefix_style>

=head2 C<log_prefix>

=head2 C<set_log_prefix>

=head2 C<is_env_debugging>

=head2 C<set_is_env_debugging>

=head2 C<into_stash>

=head2 C<set_into_stash>

=head2 C<auto_set_into>

This method any plumbing will want to call.

    $object->auto_set_into( $number_of_additional_stack_levels );

Takes a parameter to indicate the expected additional levels of stack will be need.

For instance:

    sub import {
        my ($self, %args ) = @_;
        my $object = ...->new(%args);
        $object->auto_set_into(1); # needs to be bound to the caller to import->()
    }

Or

    sub import {
        my ($self, %args ) = @_;
        my $object = ...->new(%args);
        __PACKAGE__->bar($object);

    }
    sub bar {
        $_[1]->auto_set_into(2); # skip up to caller of bar, then to caller of import
    }

And in both these cases, the end user just does:

    package::bar->import( into_level =>  0 ); # inject at this level

=head2 C<debug_prefixed_lines>

    my $code = $object->debug_prefixed_lines;
    $code->( $message );

This Debug implementation returns a C<DEBUG> sub that treats all arguments as lines of message,
and formats them as such:

    [SomePrefix::Goes::Here] this is your messages first line\n
    [SomePrefix::Goes::Here] this is your messages second line\n

The exact prefix used is determined by L<< C<log_prefix>|/log_prefix >>,
and the prefix will be omitted if C<log_prefix> is not defined.

( Note: this will likely require explicit passing of

    log_prefix => undef

)

=head2 C<debug_verbatim>

This Debug implementation returns a C<DEBUG> sub that simply
passes all parameters to C<< *STDERR->print >>, as long as debugging is turned on.

    my $code = $object->debug_verbatim;
    $code->( $message );

=head2 C<env_key_from_package>

This C<env_key_style> simply appends C<_DEBUG> to the C<env_key_prefix>

    my $key = $object->env_key_from_package;

=head2 C<env_key_prefix_from_package>

This L<< C<env_key_prefix_style>|/env_prefix_style >> converts L<< C<into>|/into >> to a useful C<%ENV> name.

    Hello::World::Bar -> HELLO_WORLD_BAR

Usage:

    my $prefix = $object->env_key_prefix_from_package;

=head2 C<log_prefix_from_package_short>

This L<< C<log_prefix_style>|/log_prefix_style >> determines a C<short> name by mutating C<into>.

When the name is C<< <10 chars >> it is passed unmodified.

Otherwise, it is tokenised, and all tokens bar the last are reduced to either

=over 4

=item a - groups of upper case only characters

=item b - failing case a, single lower case characters.

=back

    Hello -> H
    HELLO -> HELLO
    DistZilla -> DZ
    mutant -> m

And then regrouped and the last attached

    This::Is::A::Test -> T:I:A::Test
    NationalTerrorismStrikeForce::SanDiego::SportsUtilityVehicle -> NTSF:SD::SportsUtilityVehicle

Usage:

    my $prefix = $object->log_prefix_from_package_short;

=head2 C<log_prefix_from_package_long>

This L<< C<log_prefix_style>|/log_prefix_style >> simply returns C<into> as-is.

Usage:

    my $prefix = $object->log_prefix_from_package_long;

=head2 C<inject_debug_value>

Optimistically injects the desired C<$DEBUG> symbol into the package determined by C<value_name>.

Preserves the existing value if such a symbol already exists.

    $object->inject_debug_value();

=head2 C<inject_debug_sub>

Injects the desired code reference C<DEBUG> symbol into the package determined by C<sub_name>

    $object->inject_debug_sub();

=head1 ATTRIBUTES

=head2 C<debug_style>

The debug printing style to use.

    'prefixed_lines'

See L<< C<debug_styles>|/debug_styles >>

=head2 C<env_key_aliases>

A C<[]> of C<%ENV> keys that also should trigger debugging on this package.

    []

=head2 C<env_key_prefix_style>

The mechanism for determining the C<prefix> for the C<%ENV> key.

    'default'

See  L<< C<env_key_prefix_styles>|/env_key_prefix_styles >>

=head2 C<env_key_style>

The mechanism for determining the final C<%ENV> key for turning on debug.

    'default'

See L<< C<env_key_styles>|/env_key_styles >>

=head2 C<into>

The package we're injecting into.

B<IMPORTANT>: This field cannot vivify itself and be expected to work.

Because much code in this module depends on this field,
if this field is B<NOT> populated explicitly by the user, its likely
to increase the stack depth, invalidating any value if L<< C<into_level>|/into_level >> that was specified.

See L<< C<auto_set_into>|/auto_set_into >>

=head2 C<into_level>

The number of levels up to look for C<into>

Note, that this value is expected to be provided by a consuming class somewhere, and is expected to be
simply passed down from a user.

See  L<< C<auto_set_into>|/auto_set_into >> for how to set C<into> sanely.

=head2 C<sub_name>

The name of the C<CODEREF> that will be installed into C<into>

    'DEBUG'

=head2 C<value_name>

The name of the C<$SCALAR> that will be installed into C<into>

    'DEBUG' ## $DEBUG

=head2 C<env_key>

The name of the primary C<%ENV> key that controls debugging of this package.

If unspecified, will be determined by the L<< C<env_key_style>|/env_key_style >>

Usually, this will be something like

    <env_key_prefix>_DEBUG

And where C<env_key_prefix> is factory,

    <magictranslation(uc(into))>_DEBUG

Aka:

    SOME_PACKAGE_NAME_DEBUG

=head2 C<env_key_prefix>

The name of the B<PREFIX> to use for C<%ENV> keys for this package.

If unspecified, will be determined by the L<< C<env_key_prefix_style>|/env_key_prefix_style >>

Usually, this will be something like

    <magictranslation(uc(into))>

Aka:

    SOME_PACKAGE_NAME

=head2 C<debug_sub>

The actual code ref to install to do the real debugging work.

This is mostly an implementation detail, but if you were truly insane, you could pass a custom C<coderef>
to construction, and it would install the C<coderef> you passed instead of the one we generate.

Generated using L<< C<debug_style>|/debug_style >>

=head2 C<log_prefix_style>

The default style to use for C<log_prefix>.

If not set, defaults to the value of C<$ENV{PACKAGE_DEBUG_LOG_PREFIX_STYLE}> if it exists,
or simply C<'short'> if it does not.

See L<< C<log_prefix_styles>|/log_prefix_styles >>

=head2 C<log_prefix>

The string to prefix to log messages for debug implementations which use prefixes.

If not specified, will be generated from the style specified by L<< C<log_prefix_style>|/log_prefix_style >>

Which will be usually something like

    Foo::Package::Bar # 'long'
    F:P::Bar          # 'short'

=head2 C<is_env_debugging>

The determination as to whether or not the C<%ENV> indicates debugging should be enabled.

Will always be C<true> if C<$ENV{PACKAGE_DEBUG_ALL}>

And will be C<true> if either L<< C<env_key>|/env_key >> or one of L<< C<env_key_aliases>|/env_key_aliases >>
is C<true>.

B<NOTE:> This value I<BINDS> the first time it is evaluated, so for granular control of debugging at run-time,
you should not be lexically changing C<%ENV>.

Instead, you should be modifying the value of C<$My::Package::Name::DEBUG>

=head2 C<into_stash>

Contains a L<< C<Package::Stash>|Package::Stash >> object for the target package.

=head1 STYLES

=head2 C<env_key_styles>

=head3 C<default>

Uses L<< C<env_key_from_package>|/env_key_from_package >>

=head2 C<env_key_prefix_styles>

=head3 C<default>

Uses L<< C<env_key_prefix_from_package>|/env_key_prefix_from_package >>

=head2 C<log_prefix_styles>

=head3 C<short>

Uses L<< C<log_prefix_from_package_short>|/log_prefix_from_package_short >>

=head3 C<long>

Uses L<< C<log_prefix_from_package_long>|/log_prefix_from_package_long >>

=head2 C<debug_styles>

=head3 C<prefixed_lines>

Uses L<< C<debug_prefixed_lines>|/debug_prefixed_lines >>

=head3 C<verbatim>

Uses L<< C<debug_verbatim>|/debug_verbatim >>

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Package::Debug::Object",
    "interface":"class"
}


=end MetaPOD::JSON

=head1 AUTHOR

Kent Fredric <kentfredric@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4007"
  },
  "configure_requires" => {
    "Module::Build" => "0.4007"
  },
  "dist_abstract" => "Add ENV/Config controlled debug tokens to your code",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "Package-Debug",
  "dist_version" => "0.2.1",
  "license" => "perl",
  "module_name" => "Package::Debug",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Package::Stash" => "0.35",
    "perl" => "v5.10.0",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Fatal" => "0.010",
    "Test::More" => "0.98",
    "Test::Output" => "1.02"
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
